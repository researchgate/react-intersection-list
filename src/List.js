import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import warning from 'warning';
import Sentinel from './Sentinel';

const AXIS_CSS_MAP = { x: 'overflowX', y: 'overflowY' };

export default class List extends PureComponent {
    static propTypes = {
        awaitMore: PropTypes.bool,
        axis: PropTypes.oneOf(['x', 'y']),
        children: PropTypes.func,
        initialIndex: PropTypes.number,
        items(props, propName) {
            const object = props[propName];
            if (object != null && !(Array.isArray(object) || typeof object[Symbol.iterator] === 'function')) {
                return new Error(
                    `\`${propName}\` must be of type Array or a native type implementing the iterable interface`,
                );
            }
            return undefined;
        },
        itemCount: PropTypes.number,
        itemsRenderer: PropTypes.func,
        onIntersection: PropTypes.func,
        pageSize: PropTypes.number,
        renderItem: PropTypes.func,
        threshold: PropTypes.string,
    };

    static defaultProps = {
        axis: 'y',
        initialIndex: 0,
        itemsRenderer: (items, ref) => <div ref={ref}>{items}</div>,
        pageSize: 10,
        threshold: '100px',
    };

    constructor(props) {
        super(props);

        this.state = {
            size: this.computeSize(props.pageSize, this.getItemCount(props)),
        };

        this.prematureIntersectionChecked = this.state.size === 0;
    }

    setRef = callback => {
        let prevRootNode;
        this.setRootNode = node => {
            if (node !== prevRootNode) {
                prevRootNode = node;
                const overflow = window.getComputedStyle(node)[AXIS_CSS_MAP[this.props.axis]];
                callback(['auto', 'scroll', 'overlay'].indexOf(overflow) !== -1 ? node : null);
            }
        };
    };

    handleUpdate = ({ isIntersecting }) => {
        const { awaitMore, onIntersection, pageSize } = this.props;
        const { size } = this.state;

        if (!this.prematureIntersectionChecked) {
            this.prematureIntersectionChecked = true;
            warning(
                !isIntersecting,
                'ReactIntersectionList: the sentinel detected a viewport with a bigger size than the size of its items. ' +
                    'This could lead to detrimental behavior, e.g.: triggering more than one onIntersection callback at the start.\n' +
                    'To prevent this, use either a bigger `pageSize` value or avoid using the prop awaitMore initially.',
            );
        }

        if (isIntersecting) {
            const nextSize = this.computeSize(size + pageSize, this.getItemCount(this.props));
            if (this.state.size !== nextSize) {
                this.setState({ size: nextSize });
            }
            if (onIntersection && (!awaitMore || this.awaitIntersection)) {
                if (this.awaitIntersection) {
                    this.awaitIntersection = false;
                }
                onIntersection(nextSize, pageSize);
            }
        }
    };

    computeSize(pageSize, itemCount) {
        return Math.min(pageSize, itemCount);
    }

    getItemCount({ itemCount, items }) {
        const hasItemCount = typeof itemCount !== 'undefined';
        const hasItems = typeof items !== 'undefined';
        const defaultValue = 0;

        warning(
            !(hasItemCount && hasItems),
            'ReactIntersectionList: cannot use itemCount and items props at the same time, choose one to determine the number of items to render.',
        );

        if (hasItemCount) {
            return itemCount;
        }

        return hasItems ? items.length || items.size || defaultValue : defaultValue;
    }

    getItemRenderer() {
        const { children, renderItem } = this.props;
        const hasChildren = typeof children !== 'undefined';
        const hasRender = typeof renderItem !== 'undefined';

        warning(
            !(hasChildren && hasRender),
            'ReactIntersectionList: cannot use children and renderItem props as render function at the same time.',
        );

        if (hasChildren) {
            return children;
        }

        return hasRender ? renderItem : (index, key) => <div key={key}>{index}</div>;
    }

    renderItems() {
        const { awaitMore, axis, initialIndex, itemsRenderer, threshold } = this.props;
        const { size } = this.state;
        const itemRenderer = this.getItemRenderer();
        const items = [];

        for (let i = 0; i < size; ++i) {
            items.push(itemRenderer(initialIndex + i, i));
        }

        let sentinel;
        if (size < this.getItemCount(this.props) || awaitMore) {
            sentinel = (
                <Sentinel
                    key="sentinel"
                    threshold={threshold}
                    axis={axis}
                    setRef={this.setRef}
                    onChange={this.handleUpdate}
                />
            );
            items.push(sentinel);

            if (awaitMore) {
                this.awaitIntersection = true;
            }
        }

        return itemsRenderer(items, node => {
            if (node && sentinel) {
                this.setRootNode(node);
            }
        });
    }

    componentWillReceiveProps({ pageSize, ...rest }) {
        const itemCount = this.getItemCount(this.props);
        const nextItemCount = this.getItemCount(rest);

        if (this.props.pageSize !== pageSize || itemCount !== nextItemCount) {
            const nextSize = this.computeSize(this.state.size + pageSize, nextItemCount);
            this.setState({ size: nextSize });
        }
    }

    render() {
        return this.renderItems();
    }
}
