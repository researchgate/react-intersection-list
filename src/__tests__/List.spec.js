/* eslint-env jest */
import 'intersection-observer';
import React from 'react';
import renderer from 'react-test-renderer';
import List from '../';
import mockSentinel, { mockCallback } from './mockSentinel';

jest.mock('../Sentinel', () => props => {
    if (props.setRef) {
        props.setRef(mockCallback);
    }
    return mockSentinel;
});

const target = { nodeType: 1 };
const createTree = (props = {}) => renderer.create(<List {...props} />, { createNodeMock: () => target });

beforeEach(() => {
    jest.clearAllMocks();

    window.getComputedStyle = jest.fn(() => ({
        overflowX: 'auto',
        overflowY: 'auto',
    }));
});

test('render component with default props', () => {
    createTree();
});

describe('render children given pertinent props', () => {
    test('receives correct arguments given default props', () => {
        const spy = jest.fn();
        createTree({ itemCount: 10, children: spy });
        expect(spy).toHaveBeenCalledTimes(10);
        expect(spy).lastCalledWith(9, 9);
    });

    test('receives correct arguments given `pageSize` and `initialIndex` props', () => {
        const spy = jest.fn();
        createTree({
            initialIndex: 10,
            itemCount: 30,
            pageSize: 15,
            children: spy,
        });
        expect(spy).toHaveBeenCalledTimes(15);
        expect(spy).lastCalledWith(24, 14);
    });
});

describe('render component given pertinent props', () => {
    test('pure component avoids unnecessary re-rendering', () => {
        const props = { pageSize: 5, itemCount: 25 };
        const tree = createTree(props);
        const spy = jest.spyOn(tree.getInstance(), 'render');
        tree.update(<List {...props} />);
        expect(spy).not.toHaveBeenCalled();
    });

    test('throw with two different render function props', () => {
        const spy = global.spyOn(console, 'error');
        createTree({ children() {}, renderItem() {}, itemCount: 2 });
        expect(spy.calls.mostRecent().args[0]).toMatchSnapshot();
    });

    test('render children function', () => {
        const props = { children() {}, itemCount: 2 };
        const spy = jest.spyOn(props, 'children');
        createTree(props);
        expect(spy).toHaveBeenCalledTimes(2);
    });

    test('render prop function', () => {
        const props = { renderItem() {}, itemCount: 2 };
        const spy = jest.spyOn(props, 'renderItem');
        createTree(props);
        expect(spy).toHaveBeenCalledTimes(2);
    });

    test('throw with both itemCount and items props set', () => {
        const spy = global.spyOn(console, 'error');
        createTree({ items: 666 });
        expect(spy.calls.mostRecent().args[0]).toMatchSnapshot();
    });

    test('render with items instead of itemCount', () => {
        const json = createTree({ items: [1, 2] }).toJSON();
        const children = json.children;
        expect(children.length).toBe(2);
    });

    test('render empty items if no prop is given to determine length', () => {
        const instance = createTree().getInstance();
        expect(instance.getItemCount({})).toBe(0);
    });

    test('render with iterable Set', () => {
        const items = new Set([1, 2, 3]);
        const json = createTree({ items }).toJSON();
        const children = json.children;
        expect(children.length).toBe(3);
    });
});

describe('render items given sentinel', () => {
    test('sentinel observes if available items in view', () => {
        const json = createTree({ itemCount: 10, pageSize: 5 }).toJSON();
        const children = json.children;
        expect(children.length).toBe(6);
        expect(children[children.length - 1].type).toBe('span');
    });

    test('sentinel gone if no items available in view', () => {
        const json = createTree({ itemCount: 5, pageSize: 5 }).toJSON();
        const children = json.children;
        expect(children.length).toBe(5);
        expect(children[children.length - 1].type).toBe('div');
    });

    test('sentinel observes again if `itemCount` increases', () => {
        const tree = createTree({ itemCount: 5, pageSize: 5 });
        expect(tree.toJSON().children.length).toBe(5);
        tree.update(<List pageSize={5} itemCount={20} />);
        expect(tree.toJSON().children.length).toBe(11);
    });

    test('sentinel observes again if `pageSize` increases', () => {
        const tree = createTree({ itemCount: 100, pageSize: 10 });
        expect(tree.toJSON().children.length).toBe(11);
        tree.update(<List itemCount={100} pageSize={20} />);
        expect(tree.toJSON().children.length).toBe(31);
    });

    test('sentinel not present if `itemCount` increases below `pageSize`', () => {
        const tree = createTree({ itemCount: 0 });
        const children = tree.toJSON().children;
        expect(children).toBeNull();
        tree.update(<List itemCount={8} />);
        const newChildren = tree.toJSON().children;
        expect(newChildren.length).toBe(8);
        expect(newChildren[newChildren.length - 1].type).toBe('div');
    });

    test('sentinel observes with `awaitMore` bypassing the `itemCount` check', () => {
        const tree = createTree({ itemCount: 5 });
        const children = tree.toJSON().children;
        expect(children.length).toBe(5);
        tree.update(<List itemCount={5} awaitMore={true} />);
        const newChildren = tree.toJSON().children;
        expect(newChildren.length).toBe(6);
        expect(newChildren[newChildren.length - 1].type).toBe('span');
    });
});

describe('set root node', () => {
    test('ref callback sets root node', () => {
        createTree({ itemCount: 20 });
        expect(mockCallback).toBeCalledWith(target);
    });

    test('ref callback does sets root node if unmounting', () => {
        renderer.create(<List itemCount={20} />, {
            createNodeMock: () => undefined,
        });
        expect(mockCallback).not.toBeCalled();
    });

    test('ref callback does sets root node without sentinel', () => {
        const tree = createTree({ itemCount: 10 });
        expect(tree.getInstance().setRootNode).toBeUndefined();
    });

    test('ref callback sets a null root if it does not have overflow', () => {
        window.getComputedStyle = jest.fn(() => ({
            overflowY: 'visible',
        }));
        createTree({ itemCount: 20 });
        expect(mockCallback).toBeCalledWith(null);
    });

    test('having same root node prevents call getComputedStyle', () => {
        const tree = createTree({ itemCount: 20 });
        tree.getInstance().setRootNode(target);
        expect(window.getComputedStyle).toHaveBeenCalledTimes(1);
    });
});

describe('handle updates', () => {
    test('throws once if sentinel intersects with items on mount', () => {
        const spy = global.spyOn(console, 'error');
        const instance = createTree({ itemCount: 10 }).getInstance();
        instance.handleUpdate({ isIntersecting: true });
        expect(spy.calls.mostRecent().args[0]).toMatchSnapshot();
        instance.handleUpdate({ isIntersecting: true });
        expect(spy.calls.count()).toBe(1);
        createTree({ itemCount: 0 })
            .getInstance()
            .handleUpdate({ isIntersecting: true });
        expect(spy.calls.count()).toBe(1);
    });

    test('sets next size value computed into `pageSize`', () => {
        const instance = createTree({ itemCount: 20 }).getInstance();
        instance.handleUpdate({ isIntersecting: false });
        instance.handleUpdate({ isIntersecting: true });
        expect(instance.state.size).toBe(20);
    });

    test('sets next size value computed into `itemCount`', () => {
        const instance = createTree({ itemCount: 15 }).getInstance();
        instance.handleUpdate({ isIntersecting: false });
        instance.handleUpdate({ isIntersecting: true });
        expect(instance.state.size).toBe(15);
    });

    test('invokes `onIntersection` each time when `awaitIntersection` is false', () => {
        const spy = jest.fn();
        const instance = createTree({
            itemCount: 30,
            onIntersection: spy,
        }).getInstance();
        instance.handleUpdate({ isIntersecting: false });
        instance.handleUpdate({ isIntersecting: true });
        instance.handleUpdate({ isIntersecting: true });
        expect(instance.state.size).toBe(30);
        expect(spy).toHaveBeenCalledTimes(2);
    });

    test('invokes `onIntersection` only once when `awaitIntersection` is true', () => {
        const spy = jest.fn();
        const props = {
            awaitMore: true,
            itemCount: 10,
            onIntersection: spy,
        };
        const tree = createTree(props);
        tree.getInstance().handleUpdate({ isIntersecting: false });
        tree.getInstance().handleUpdate({ isIntersecting: true });
        tree.getInstance().handleUpdate({ isIntersecting: true });
        expect(tree.getInstance().state.size).toBe(10);
        tree.update(<List {...props} itemCount={20} />);
        tree.getInstance().handleUpdate({ isIntersecting: true });
        expect(spy).toHaveBeenCalledTimes(2);
    });
});
