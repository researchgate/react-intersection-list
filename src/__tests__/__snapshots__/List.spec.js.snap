// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`handle updates throws once if sentinel intersects with items on mount 1`] = `
"Warning: ReactIntersectionList: the sentinel detected a viewport with a bigger size than the size of its items. This could lead to detrimental behavior, e.g.: triggering more than one onIntersection callback at the start.
To prevent this, use either a bigger \`pageSize\` value or avoid using the prop awaitMore initially."
`;

exports[`render component given pertinent props throw with both itemCount and items props set 1`] = `
"Warning: Failed prop type: \`items\` must be of type Array or a native type implementing the iterable interface
    in List"
`;

exports[`render component given pertinent props throw with two different render function props 1`] = `"Warning: ReactIntersectionList: cannot use children and renderItem props as render function at the same time."`;
